include/geojson/features/FeatureBase.hpp

```diff
@@ -0,0 +1,471 @@
+                    }
+                }
+
+                return longest + 1;
+            }
+
+            int get_contributor_count() {
+                int count = 0;
+
+                std::vector<FeatureBase*> to_search;
+                std::vector<FeatureBase*> total;
+
+                FeatureBase *current = this;
+                total.push_back(current);
+
+                for (FeatureBase *upstream_feature : current->upstream) {
+                    to_search.push_back(upstream_feature);
+                }
+
+                while (to_search.size() > 0) {
+                    current = to_search.back();
+                    to_search.pop_back();
+                    total.push_back(current);
+                    for (FeatureBase *upstream_feature : current->upstream) {
+                        if (not contains(total, upstream_feature) and not contains(to_search, upstream_feature)) {
+                            to_search.push_back(upstream_feature);
+                        }
+                    }
+                }
+
+                return total.size() - 1;
+            }
+
+            bool is_leaf() {
+                return this->downstream.size() == 0;
+            }
+
+            bool is_root() {
+                return this->upstream.size() == 0;
+            }
+
+            template<class T>
+            T geometry() const {
+                try {
+                    return boost::get<T>(this->geom);
+                }
+                catch (boost::bad_get exception) {
+                    std::string template_name = boost::typeindex::type_id<T>().pretty_name();
+                    std::string expected_name = get_geometry_type(this->geom);
+                    std::cerr << "Asked for " << template_name << ", but only " << expected_name << " is valid" << std::endl;
+                    throw;
+                }
+            }
+
+            ::geojson::geometry geometry();
+
+            template<class T>
+            T geometry(int index) const {
+                try {
+                    return boost::get<T>(this->geometry_collection[index]);
+                }
+                catch (boost::bad_get exception) {
+                    std::string template_name = boost::typeindex::type_id<T>().pretty_name();
+                    std::string expected_name = get_geometry_type(this->geometry_collection[index]);
+                    std::cerr << "Asked for " << template_name << ", but only " << expected_name << " is valid" << std::endl;
+                    throw;
+                }
+            }
+
+            virtual void visit(FeatureVisitor &visitor) = 0;
+        protected:
+            FeatureType type;
+            ::geojson::geometry geom;
+            std::vector<::geojson::geometry> geometry_collection;
+
+            PropertyMap properties;
+            std::vector<double> bounding_box;
+            PropertyMap foreign_members;
+            std::string id;
+
+            std::vector<FeatureBase*> upstream;
```

Comment:
Are upstream and downstream ordered in the vector (i.e. downstream[0] is first downstream, downstream[1] is two hops away?  If not, a map might be a better alternative, the lookups in get_(up/down)stream would be much quicker.  If order is important though, the vector might be the best.

Reply:
The order isn't really important; if you want to get a feature that's two hops away, you'd go to downstream[0]->downstream[0].

Reply:
So its link-list like.  a map of id -> feature would be appropriate.